### ISSUE TITLE
Controllable DelayQueue / Timer wheel

### ISSUE THREAD

**Issue by Diggsey** [tokio-rs/tokio#2719]
I'm building a kind of "test-bed" for some tokio code that uses timers. I would like to be able to not only "pause time" for this tokio code, but also to adjust the rate at which time passes, so I can observe the code at eg. 1/1000th of its normal speed.

Tokio already has some timer test functionality, but it's not really designed for this use-case. Ideally there would be a way to create a handle to a timer wheel which I can pass into the tokio code, and then later pause and/or adjust the speed multiplier of the timer wheel.

I can then pass this handle into code running in the test-bed, and there would be versions of methods like `delay_for` that take a handle, which the algorithm can use.

**Comment by carllerche** [comment]
I'm thinking `DelayQueue` should just be the structure and not internally check the clock. The caller would advance the time however they see fit. There probably would be some utility fns that do check `Instant::now()` but they would be an additional layer.

Is that roughly what you mean?

**Comment by Diggsey** [comment]
Yeah - I think I could build what's necessary out of that if there was a "time to next event" function on the delay queue.

IIUC, I'd need to select over `delay_for(queue.time_to_next() / speed_multiplier)` and a channel on which to receive updates to the `speed_multiplier` variable. On each wake-up, I'd manually advance the queue by `actual_elapsed_time * speed_multiplier`. Does that sound like it would work?

**Comment by carllerche** [comment]
@Diggsey Yep, that should work fine.

**Comment by Diggsey** [comment]
@carllerche the other thing that would be useful, but not sure how feasible it is, is to be able to abstract over this custom delay queue and the "built in" delay queue that tokio uses.
